### ISSUE TITLE
Jacobian.backward costs a lot of time

### ISSUE THREAD

**Issue by santisy** [torch/nn#820]
I have a neural network and want to get the Jacobian matrix of its outputs wrt. its parameters. Here is my code
```lua
jac = nn.Jacobian
para, dpara = uModel:getParameters()
uJacobian = jac.backward(uModel, inputs, para, dpara)
```
uModel here is a convolutional neural network containing batchnormalization. It has 14841291
learnable parameters and here inputs is (16, 3, 60, 60) tensor, and output is (16, 3) tensor. All data are in GPU. 
According to the source code of `nn.Jacobian`, I think it only dose 16x3=48 loops to compute gradients to parameters, but the process above cost about 65 seconds. My computer has Intel® Xeon(R) CPU E5-2620 v3 @ 2.40GHz × 24  and Tesla k20c to do this computing. And I think it has done it a little bit slow. Did I do something wrong here? How can I improve the performance here?

**Comment by fmassa** [comment]
`nn.Jacobian` is only used for testing the correctness of a forward/backward implementation. It indeed requires running forward multiple times and doing finite differences to compute the gradient, which becomes super slow when the dimension of the input increases (as it increases the number of forward passes to run).

**Comment by santisy** [comment]
@fmassa thanks for reply, so any suggestion about how can I implement the calculation of Jacobian Matrix of outputs of a neural network with respect to the parameters efficiently?

**Comment by RuiShu** [comment]
It looks like `nn.Jacobian.backward` does in fact use backpropagation, not finite differences. `module:forward(input)` is only called once, but what kills the speed is the fact that `module:accGradParameters(input, dout)` is called multiple times. To the best of my knowledge, unfortunately, this is the most efficient way of computing the Jacobian.

**Comment by santisy** [comment]
@RuiShu I think so. But the time is really unusual too long. I use my Tesla k20c to compute 100 backwards of a neural network with similar scale and only cost about only 9 seconds. But here it costs almost 66 seconds. Does it owe to the operation `.new()` and we formerly has all data in GPU? Is there transferring time or something like that kills the time?

**Comment by RuiShu** [comment]
`module.output.new()` simply creates a new Tensor object whose type (float, double, cuda, etc) matches that of the module. It's a very quick and painless operation and is unlikely to be the problem.

**Comment by RuiShu** [comment]
Out of curiosity, what's the size of your output? (i.e. the value for `uModel:forward(input):nElement()`. Is it 100?)

**Comment by santisy** [comment]
@RuiShu 48 = 3*16. 3 dimensions of output and 16 mini-batch size.

**Comment by santisy** [comment]
@RuiShu I think I have found the reason, if you directly call `Jacobian.backward` it somehow not call the function using backpropagation but another function that has the same name, maybe that function actually using finite differences to compute gradients.

**Comment by RuiShu** [comment]
`nn.Jacobian.backward` works as expected for me.
